dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/sms/job/InboundSmsProcessingJob.java

```diff
@@ -25,72 +25,70 @@
  * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
  * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
  */
-package org.hisp.dhis.sms;
+package org.hisp.dhis.sms.job;
 
 import java.util.List;
 import lombok.RequiredArgsConstructor;
-import lombok.extern.slf4j.Slf4j;
 import org.hisp.dhis.message.MessageSender;
+import org.hisp.dhis.scheduling.Job;
+import org.hisp.dhis.scheduling.JobConfiguration;
+import org.hisp.dhis.scheduling.JobProgress;
+import org.hisp.dhis.scheduling.JobType;
+import org.hisp.dhis.scheduling.parameters.SmsInboundProcessingJobParameters;
 import org.hisp.dhis.sms.incoming.IncomingSms;
 import org.hisp.dhis.sms.incoming.IncomingSmsListener;
 import org.hisp.dhis.sms.incoming.IncomingSmsService;
 import org.hisp.dhis.sms.incoming.SmsMessageStatus;
-import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.beans.factory.annotation.Qualifier;
 import org.springframework.stereotype.Component;
 
-@Slf4j
+@Component
 @RequiredArgsConstructor
-@Component("org.hisp.dhis.sms.SmsConsumerThread")
-public class SmsConsumerThread {
-  private List<IncomingSmsListener> listeners;
+public class InboundSmsProcessingJob implements Job {
+  private final IncomingSmsService incomingSmsService;
 
-  private final MessageQueue messageQueue;
+  private final List<IncomingSmsListener> listeners;
 
   @Qualifier("smsMessageSender")
   private final MessageSender smsSender;
 
-  private final IncomingSmsService incomingSmsService;
+  @Override
+  public JobType getJobType() {
+    return JobType.SMS_INBOUND_PROCESSING;
+  }
 
-  public void spawnSmsConsumer() {
-    IncomingSms message = messageQueue.get();
+  @Override
+  public void execute(JobConfiguration config, JobProgress progress) {
+    SmsInboundProcessingJobParameters params =
+        (SmsInboundProcessingJobParameters) config.getJobParameters();
+    progress.startingProcess("Process incoming SMS with id {}", params.getSmsId());
 
-    while (message != null) {
-      log.info("Received SMS: " + message.getText());
+    IncomingSms sms = incomingSmsService.get(params.getSmsId());
+    if (sms == null) {
+      progress.failedProcess("No incoming SMS found for id {}", params.getSmsId());
+      return;
+    }
```

Comment:
Usually I would advice on using stages in the process to execute the steps. But since this is very basic and we might want to keep it light weight it may be a good case where process level tracking is all we need. 

If you want to use stage level the work like `incomingSmsService.get` should be run via `runStage`. Look at usages of `JobProgress.nonNullStagePostCondition`. First the service all gets run via `runStage`. Then the `nonNullStagePostCondition` makes sure the result is non-null. This would happen as any exception in `runStage` would be caught and `null` would be returned. Wrapping that in `nonNullStagePostCondition` makes sure such issues fail eagerly as opposed to lazily at the next `startingStage` call. Obviously when you make the loading a stage you also need an `startingStage` before the `runStage` part.

If you decide to use stages the second part of finding and receiving via listener would be another stage.

Reply:
That sounds to complicated for me right now 😅